" " "
RNASeq workflow with FASTQ files as input and a count table as output
Nghi Nguyen
" " "

BASE_DIR ="/fs/ess/PAS1855/users/nghinguyen/FinalProject/Snakemake"

REF_FA = WDIR + "refgenome/Gmax_275_v2.0.fa"

SAMPLES, = glob_wildcards(BASE_DIR + "/{sample}_1.fastq")

READS=["1", "2"]


rule all:
    """
    input:
        "res/count_table.txt",
         expand("res/{sample}_{read}.fastqc.html", sample=SAMPLES, read=READS)

rule trim:
    input:
        "data/{SRR}.fastq",
    output:
        "res/{SRR}_trim.fastq",
    log:
        "log/trim/{SRR}.log",
    shell:
        "scripts/trim.sh {input} >{output} 2>{log}"


rule map:
    """
    Map the trimmed FASTQ files to a reference genome.
    """
    input:
        R1=data/{sample}_1.fastq 
        R2=data/{sample}_2.fastq
        ref_dir=REF_FA,
        out_prefix=?
    output:
        "res/{SRR}.bam",
log:
        "log/map/{SRR}.log",
  shell:
  shell: star_submit.sh {input.R1} {input.R2} {input.refdir} {input.out_prefix}

rule count:
    """
    From the BAM files, create a gene count table.
    """
    input:
        expand("res/{SRR}.bam", SRR=SAMPLES),
    output:
        "res/count_table.txt",
      log:
        "log/count/count.log",
    shell:
        "scripts/count.sh {input} >{output} 2>{log}"

rule fastqc:
    input:
        "data/{SRR}.fastq",
    output:
        "res/{SRR}.fastqc.html",
    log:
        "log/fastqc/{SRR}.log",
    shell:
        "scripts/fastqc.sh {input} res &>{log}"

